### DISCUSSION TITLE
Currify enum constructors

### DISCUSSION THREAD

**Discussion by mlutze** [flix/flix#6745] [Ideas]
Enum constructors currently roughly use the OCaml style: n-ary constructors take an n-ary tuple as an argument.

I propose instead that we adopt the Haskell style: n-ary constructors act as n-ary functions: they become curryable.

We already have a slight hack in the compiler to support this behavior for 0-ary constructors: When I write `None`, which is a `Unit -> Option[a]` constructor, the compiler sees it is a `Unit -> _` constructor and treats this simply as an `Option[a]`. On the other hand, if I write `Some`, the compiler sees it as a `a -> Option[a]` constructor and keeps it as a function.

The Haskell style would mean more consistency in the language.

**Comment by magnus-madsen** [comment]
What about pattern matching?

**Reply by mlutze** [reply]
Pattern matching would work the same as it currently does.
